frontend/src/context/PlaygroundProvider.tsx

```diff
@@ -33,20 +43,75 @@ export const PlaygroundProvider: React.FC<{ children: React.ReactNode }> = ({
     }
   }, []);
 
+  // Keep track of the previous valid state that was successfully saved
+  const [lastValidState, setLastValidState] = useState<{
+    inputDocuments: InputDocument[];
+    transformations: Transformation[];
+  }>({ inputDocuments: [], transformations: [] });
+
   // Save state to localStorage whenever it changes (excluding outputDocuments)
   useEffect(() => {
-    try {
-      const stateToSave = {
-        inputDocuments: state.inputDocuments,
-        transformations: state.transformations,
-      };
-      localStorage.setItem(STORAGE_KEY, JSON.stringify(stateToSave));
-    } catch (error) {
-      console.error("Failed to save state to localStorage:", error);
+    // Skip the effect if we're in the process of reverting to the last valid state
+    const stateToSave = {
+      inputDocuments: state.inputDocuments,
+      transformations: state.transformations,
+    };
+
+    const stateJson = JSON.stringify(stateToSave);
+
+    // Update the total storage size
+    const sizeInBytes = new Blob([stateJson]).size;
+    setTotalStorageSize(sizeInBytes);
```

Comment:
I was a bit confused why we were calculating size and storing in both effect functions. Do we just not store this value in state?

Reply:
We do store it in state, but not in local storage.
The first `useEffect` (lines 25-44) is only run when the provider first renders, so it calculates how much space we're currently using from whatever is loaded from local storage.

The second `useEffect` (lines 53-94) runs when inputDocuments or transformations changes, and it theoretically could use the saved `storageSize` and just modify it by whatever the change was (e.g. add the size of the latest document), but it's implemented such that it doesn't know what the change was (e.g. which document was added, updated, or removed), it just has the new state in its entirety.